### PR TITLE
Throw exception when schema name doesn't match table name during table creation

### PR DESCRIPTION
**Pull request by xiangfu0**
Throw exception when schema doesn't match table name

### CODE REVIEW

**Comment by Jackie-Jiang** [comment]
cc @mcvsubbu @siddharthteotia @yupeng9 @jadami10 
We already deprecated the schema name from table config for years (since 2019), and suggest using raw table name as schema name for the following reasons:
- Error prune: sometimes user accidentally copy the table config and point to the wrong schema
- Easier lookup: we can easily look up the schema for a table without going 2 hops, more importantly, we always know the table associated with the schema. Currently we need to scan all the tables in order to find if a schema can be deleted
- Easier management: in the future we can consider managing table config and schema together (we already started this by introducing the `TableConfigs` concept

Now we want to start the process of enforcing it. Do you have concern about this?
After preventing new table with different schema being created, we will provide an one time migration to rename the schema to match the table config.

**Review by mcvsubbu — COMMENTED** [review]
It looks like this will cause a problem for us, since some of the users are sending a specific string to indicate that the schema does not exist.

We will dig a little bit more and get back to you.

**Comment by mcvsubbu** [comment]
So, am I correct that this PR only prevents new tables from getting created without a proper schema name. Existing tables, if they are different schema name, will continue to work as they do now?

I suppose at some point, that will be removed as well, right? (I see code everywhere that corrects the situation for a schema not being present in the same name as the tableName).

**Comment by mcvsubbu** [comment]
cc: @GSharayu , @sajjad-moradi

**Comment by Jackie-Jiang** [comment]
@mcvsubbu 

> So, am I correct that this PR only prevents new tables from getting created without a proper schema name. Existing tables, if they are different schema name, will continue to work as they do now?

Yes, this PR only prevents new tables with different schema name to be created. For the existing tables, we will do a one time migration later to move/copy the schema if they do not match the table name

> I suppose at some point, that will be removed as well, right? (I see code everywhere that corrects the situation for a schema not being present in the same name as the tableName).

You mean the schema name field? Once we make all the schema match the table name, we can remove that field

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/UpsertTableIntegrationTest.java [hunk 1]
```diff
@@ -245,14 +244,13 @@ protected void testDeleteWithFullUpsert(String kafkaTopicName, String tableName,
...
 
-
     // Validate: pk is queryable and all columns are overwritten with new value
     rs = getPinotConnection()
         .execute("SELECT playerId, name, game FROM " + tableName + " WHERE playerId = 100").getResultSet(0);
     Assert.assertEquals(rs.getRowCount(), 1);
     Assert.assertEquals(rs.getInt(0, 0), 100);
     Assert.assertEquals(rs.getString(0, 1), "Zook-New");
-    Assert.assertEquals(rs.getString(0, 2), "null");
+    Assert.assertEquals(rs.getString(0, 2), "[\"null\"]");
```
Comment (xiangfu0) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/UpsertTableIntegrationTest.java:253]: @shenyu0127 is this behavior change expected?

Comment (shenyu0127) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/UpsertTableIntegrationTest.java:253]: This behavior is not expected.

Your changes to `PinotHelixResourceManager.java` are unrelated to ingestion. Why would the behavior change?

**Comment by xiangfu0** [comment]
> Hi @xiangfu0 @Jackie-Jiang ,
> 
> As @mcvsubbu mentioned, We are working with partner team to resolve the table creation flow since this will error out for new tables. As I understand from the discussion, this will be followed by a script to apply the changes to the existing tables. We will also need to work with the team for the same. We can get back to you on this PR pretty soon with any updates

Yes, plan to add a periodic task to fix the schema name for existing tables: https://github.com/apache/pinot/pull/11574

This will prevent new table creation.

**Comment by yupeng9** [comment]
yes, this is a good move.

**Comment by xiangfu0** [comment]
> Hi @xiangfu0 @Jackie-Jiang ,
> 
> As @mcvsubbu mentioned, We are working with partner team to resolve the table creation flow since this will error out for new tables. As I understand from the discussion, this will be followed by a script to apply the changes to the existing tables. We will also need to work with the team for the same. We can get back to you on this PR pretty soon with any updates

Do you have eta for this?

**Comment by GSharayu** [comment]
> > Hi @xiangfu0 @Jackie-Jiang ,
> > As @mcvsubbu mentioned, We are working with partner team to resolve the table creation flow since this will error out for new tables. As I understand from the discussion, this will be followed by a script to apply the changes to the existing tables. We will also need to work with the team for the same. We can get back to you on this PR pretty soon with any updates
> 
> Do you have eta for this?

Hi @xiangfu0 , the partner team provided a time estimate of one week.
cc @mcvsubbu

**Comment by xiangfu0** [comment]
> > > Hi @xiangfu0 @Jackie-Jiang ,
> > > As @mcvsubbu mentioned, We are working with partner team to resolve the table creation flow since this will error out for new tables. As I understand from the discussion, this will be followed by a script to apply the changes to the existing tables. We will also need to work with the team for the same. We can get back to you on this PR pretty soon with any updates
> > 
> > 
> > Do you have eta for this?
> 
> Hi @xiangfu0 , the partner team provided a time estimate of one week. cc @mcvsubbu

Sounds good, we can merge this by the EOM.

**Comment by mcvsubbu** [comment]
I have a more general question on this PR. Should this PR not include places in pinot-spi where we can throw exception if a client tries to create a table config that does not have a schema with the same name? Right now, the `TableConfigBuilder` class accepts any schema as the schemaName.  We should be throwing an exception there, so that these problems are caught earlier?

**Comment by xiangfu0** [comment]
> I have a more general question on this PR. Should this PR not include places in pinot-spi where we can throw exception if a client tries to create a table config that does not have a schema with the same name? Right now, the `TableConfigBuilder` class accepts any schema as the schemaName. We should be throwing an exception there, so that these problems are caught earlier?

Most of the users are using REST endpoint to POST a json table config to create table.
It's ok to catch earlier in TableConfigBuilder, just the validation cannot be ignored from the server side.

**Comment by mcvsubbu** [comment]
> > I have a more general question on this PR. Should this PR not include places in pinot-spi where we can throw exception if a client tries to create a table config that does not have a schema with the same name? Right now, the `TableConfigBuilder` class accepts any schema as the schemaName. We should be throwing an exception there, so that these problems are caught earlier?
> 
> Most of the users are using REST endpoint to POST a json table config to create table. It's ok to catch earlier in TableConfigBuilder, just the validation cannot be ignored from the server side.

In that case, can we please include the change in spi as well? thanks.

We ask our partners to use spi, since that has a better guarantee of keeping backward compat (that is the purpose of spi, right?)

**Comment by xiangfu0** [comment]
> > > I have a more general question on this PR. Should this PR not include places in pinot-spi where we can throw exception if a client tries to create a table config that does not have a schema with the same name? Right now, the `TableConfigBuilder` class accepts any schema as the schemaName. We should be throwing an exception there, so that these problems are caught earlier?
> > 
> > 
> > Most of the users are using REST endpoint to POST a json table config to create table. It's ok to catch earlier in TableConfigBuilder, just the validation cannot be ignored from the server side.
> 
> In that case, can we please include the change in spi as well? thanks.
> 
> We ask our partners to use spi, since that has a better guarantee of keeping backward compat (that is the purpose of spi, right?)

Removed the `setSchemaName` method from `TableConfigBuilder`

**Comment by xiangfu0** [comment]
Since we are already at the month end. Let's merge this?
@Jackie-Jiang @mcvsubbu @GSharayu

**Comment by Jackie-Jiang** [comment]
@mcvsubbu @GSharayu Let us know if we can go ahead and merge this. Once it is merged, the new created table must have schema, and schema name must match raw table name.

**Comment by GSharayu** [comment]
Hi @xiangfu0, feel free to merge the PR.  The changes for new tables have been completed by our partner teams. Thanks a lot!

**Comment by xiangfu0** [comment]
> Hi @xiangfu0 and @Jackie-Jiang , feel free to merge the PR. The changes for new tables have been completed by our partner teams. Thanks a lot!

Thanks!

**Pull request merged**